src/kit/shell/src/shellEngine.c

```diff
@@ -1362,45 +1366,60 @@ int wsclient_conn() {
   return -1;
 }
 
-cJSON *wsclient_parse_response() {
-  char *recv_buffer = calloc(1, 4096);
-  int   start = 0;
-  bool  found = false;
-  int   received = 0;
-  int   bytes;
-  int   recv_length = 4095;
-  do {
-    bytes = recv(args.socket, recv_buffer + received, recv_length - received, 0);
-    if (bytes == -1) {
-      free(recv_buffer);
-      fprintf(stderr, "websocket recv failed with bytes: %d\n", bytes);
-      return NULL;
-    }
-
-    if (!found) {
-      for (; start < recv_length - received; start++) {
-        if ((recv_buffer + start)[0] == '{') {
-          found = true;
-          break;
-        }
-      }
+void wsclient_parse_frame(SWSParser * parser, uint8_t * recv_buffer) {
+  unsigned char msg_masked = (recv_buffer[1] >> 7) & 0x01;
+  int payload_length = 0;
+  int pos = 2;
+  int length_field = recv_buffer[1] &(~0x80);
+  unsigned int mask = 0;
+  if (length_field <= 125) {
+    payload_length = length_field;
+  } else if (length_field == 126) {
+    payload_length = recv_buffer[2];
+    for (int i = 0; i < 1; i++) {
+      payload_length = (payload_length << 8) + recv_buffer[3 + i];
     }
-    if (NULL != strstr(recv_buffer + start, "}")) {
-      break;
+    pos += 2;
+  } else if (length_field == 127) {
+    payload_length = recv_buffer[2];
+    for (int i = 0; i < 7; i++) {
+      payload_length = (payload_length << 8) + recv_buffer[3 + i];
     }
-    received += bytes;
-    if (received >= recv_length) {
-      recv_length += 4096;
-      recv_buffer = realloc(recv_buffer + start, recv_length);
+    pos += 8;
+  }
+  if (msg_masked) {
+    mask = *((unsigned int *) (recv_buffer + pos));
+    pos += 4;
+    const uint8_t *c = recv_buffer + pos;
+    for (int i = 0; i < payload_length; i++) {
+      recv_buffer[i] = c[i] ^ ((unsigned char *) (&mask))[i % 4];
     }
-  } while (1);
-  cJSON *res = cJSON_Parse(recv_buffer + start);
-  if (res == NULL) {
-    fprintf(stderr, "fail to parse response into json: %s\n", recv_buffer + start);
+  }
+  parser->offset = pos;
+  parser->payload_length = payload_length;
+}
+
+char *wsclient_get_response() {
+  uint8_t *recv_buffer = calloc(1, 1024);
```

Comment:
How about putting it on the stack as an array of just 1024 bytes if you don't want to check allocation succeed?